Add Barbican adoption with Proteccio HSM integration documentation (#1058)
openstack-k8s-operators/data-plane-adoption: docs_user/modules/ref_troubleshooting-key-manager-hsm-adoption.adoc

= Troubleshooting Key Manager HSM adoption

[id="troubleshooting-key-manager-hsm-adoption_{context}"]

This section provides troubleshooting guidance for common issues encountered during HSM-enabled Key Manager (Barbican) service adoption.

== Configuration validation failures

.Problem
The adoption fails with validation errors about placeholder values.

.Symptoms
----
TASK [Validate all required variables are set] ****
fatal: [localhost]: FAILED! => {
    "msg": "Required variable proteccio_certs_path contains placeholder value."
}

> ----

Reviewer: ```suggestion

```
I moved the content to the introduction.

> .Solution

Reviewer: ```suggestion
.Procedure
```

> . Edit your HSM configuration in the Zuul job vars or CI framework configuration file.

Reviewer: ```suggestion
. Edit your hardware security module configuration in the Zuul job vars or CI framework configuration file.
```

> . Replace all placeholder values with actual configuration values for your environment. Key variables to check:

Reviewer: ```suggestion
. Replace all placeholder values with actual configuration values for your environment. Check the following key variables:
```

+
[source,yaml]
----
cifmw_hsm_password: "your_actual_hsm_password"
cifmw_barbican_proteccio_partition: "VHSM1"
cifmw_barbican_proteccio_mkek_label: "your_mkek_label"
cifmw_barbican_proteccio_hmac_label: "your_hmac_label"
cifmw_hsm_proteccio_client_src: "https://your-server/path/to/Proteccio.iso"

> cifmw_hsm_proteccio_conf_src: "https://your-server/path/to/proteccio.rc"

Reviewer: ```suggestion
cifmw_hsm_password: <your_actual_hsm_password>
cifmw_barbican_proteccio_partition: <VHSM1>
cifmw_barbican_proteccio_mkek_label: <your_mkek_label>
cifmw_barbican_proteccio_hmac_label: <your_hmac_label>
cifmw_hsm_proteccio_client_src: <https://your-server/path/to/Proteccio.iso>
cifmw_hsm_proteccio_conf_src: <https://your-server/path/to/proteccio.rc>
```

----

. Verify no placeholder values remain in your configuration.

== HSM file prerequisites missing

.Problem
The adoption fails because HSM certificates or client software cannot be found.

.Symptoms
----
TASK [Validate Proteccio prerequisites exist] ****
fatal: [localhost]: FAILED! => {
    "msg": "Proteccio client ISO not found: /opt/proteccio/Proteccio3.06.05.iso"
}

> ----

Reviewer: ```suggestion
[role="_abstract"]
If the adoption fails because HSM certificates or client software cannot be found, update your configuration to point to the files in their specific locations. The error looks similar to the following example:
+
----
TASK [Validate Proteccio prerequisites exist] ****
fatal: [localhost]: FAILED! => {
    "msg": "Proteccio client ISO not found: /opt/proteccio/Proteccio3.06.05.iso"
}
----
```

.Solution
. Verify all required HSM files are accessible from the configured URLs:
+
[source,bash]
----
$ curl -I https://your-server/path/to/Proteccio3.06.05.iso
$ curl -I https://your-server/path/to/proteccio.rc
$ curl -I https://your-server/path/to/client.crt
$ curl -I https://your-server/path/to/client.key
----

. If files are in different locations, update the URL variables in your configuration:
+
[source,yaml]
----
cifmw_hsm_proteccio_client_src: "https://correct-server/path/to/Proteccio3.06.05.iso"
cifmw_hsm_proteccio_conf_src: "https://correct-server/path/to/proteccio.rc"
cifmw_hsm_proteccio_client_crt_src: "https://correct-server/path/to/client.crt"
cifmw_hsm_proteccio_client_key_src: "https://correct-server/path/to/client.key"
----

> . Ensure the URLs are accessible from the CI environment (check network connectivity and authentication)

Reviewer: ```suggestion
. Check the network connectivity and authentication to ensure that the URLs are accessible from the CI environment.
```
Please make sure my rewrite is accurate.

    "msg": "SSH connection to source environment failed"
}
----

.Solution
. Verify SSH connectivity to the source controller:
+
[source,bash]
----
$ ssh -o StrictHostKeyChecking=no tripleo-admin@controller-0.ctlplane
----

. Update the `controller1_ssh` variable if needed:
+
[source,yaml]
----
controller1_ssh: "ssh -o StrictHostKeyChecking=no tripleo-admin@<controller_ip>"
----

. Ensure SSH keys are properly configured for passwordless access

== HSM secret creation failures

.Problem
HSM secrets cannot be created in the target environment.

.Symptoms
----
TASK [Create HSM secrets in target environment] ****
fatal: [localhost]: FAILED! => {
    "msg": "Failed to create secret proteccio-data"
}
----

.Solution
. Verify target environment access:
+
[source,bash]
----
$ export KUBECONFIG=/path/to/.kube/config
$ oc get secrets -n openstack
----

. Check if secrets already exist:
+
[source,bash]
----
$ oc get secret proteccio-data hsm-login -n openstack
----

. If secrets exist with different names, update the configuration variables:
+
[source,yaml]
----
proteccio_login_secret_name: "your-hsm-login-secret"
proteccio_client_data_secret_name: "your-proteccio-data-secret"
----

> 

Reviewer: What if the secrets don't exist? How do customers create one?

$ podman tag hello-world <registry>/<namespace>/test:latest
$ podman push <registry>/<namespace>/test:latest
----

. Update registry configuration variables if needed:
+
[source,yaml]
----
cifmw_update_containers_registry: "your-registry:5001"
cifmw_update_containers_org: "your-namespace"
cifmw_image_registry_verify_tls: false
----

== HSM backend detection failures

.Problem
The adoption role cannot detect HSM configuration in the source environment.

.Symptoms
----
TASK [detect source environment HSM configuration] ****
ok: [localhost] => {
    "msg": "No HSM configuration found - using standard adoption"
}
----

.Solution
. Manually verify HSM configuration exists in the source:
+
[source,bash]
----
$ ssh tripleo-admin@controller-0.ctlplane \
  "sudo grep -A 10 '\[p11_crypto_plugin\]' \
  /var/lib/config-data/puppet-generated/barbican/etc/barbican/barbican.conf"
----

. If HSM is configured but not detected, force HSM adoption by setting the `barbican_hsm_enabled` variable:
+
[source,yaml]
----
# In your Zuul job vars or CI framework configuration
barbican_hsm_enabled: true
----
+
This will ensure the `barbican_adoption` role uses the HSM-enabled patch for Barbican deployment.

> == Database migration issues

Reviewer: ```suggestion
== Resolving database migration issues
```

.Problem
HSM metadata is not preserved during database migration.

.Symptoms
----
TASK [Verify database migration preserves HSM references] ****
ok: [localhost] => {
    "msg": "HSM secrets found in migrated database: 0"
}
----

.Solution
. Verify source database contains HSM secrets:
+
[source,bash]
----
$ ssh tripleo-admin@controller-0.ctlplane \
  "sudo mysql barbican -e 'SELECT COUNT(*) FROM secret_store_metadata WHERE key=\"plugin_name\" AND value=\"PKCS11\";'"
----

. Check database migration logs for errors:
+
[source,bash]
----
$ oc logs deployment/barbican-api | grep -i migration
----

> . If migration failed, restore from backup and retry

Reviewer: ```suggestion
. If the migration failed, restore the database from backup and retry.
```
Can we give any general guidance on how to do this, or is this basic knowledge?

== Service startup failures

.Problem
Barbican services fail to start after HSM configuration is applied.

.Symptoms
----
$ oc get pods -l service=barbican
NAME                           READY   STATUS    RESTARTS   AGE
barbican-api-xyz               0/1     Error     0          2m
----

> .Solution

Reviewer: ```suggestion
== Resolving service startup failures

[role="_abstract"]
If the {key_manager_first_ref} services fail to start after the hardware security module (HSM) configuration is applied, check the configuration in the pod. The error for the services failing to start looks similar to the following example:
+
----
$ oc get pods -l service=barbican
NAME                           READY   STATUS    RESTARTS   AGE
barbican-api-xyz               0/1     Error     0          2m
----

.Procedure
```

. Check pod logs for HSM connectivity issues:
+
[source,bash]
----
$ oc logs barbican-api-xyz
----

. Verify HSM library is accessible:
+
[source,bash]
----
$ oc exec barbican-api-xyz -- ls -la /usr/lib64/libnethsm.so
----

. Check HSM configuration in the pod:
+
[source,bash]
----
$ oc exec barbican-api-xyz -- cat /etc/proteccio/proteccio.rc
----

> == Performance and connectivity issues

Reviewer: ```suggestion
== Resolving performance and connectivity issues
```

.Problem
HSM operations are slow or fail intermittently.

.Solution
. Test HSM connectivity from Barbican pods:
+
[source,bash]
----
$ oc exec barbican-api-xyz -- pkcs11-tool --module /usr/lib64/libnethsm.so --list-slots
----

. Check HSM server connectivity:
+
[source,bash]
----
$ oc exec barbican-api-xyz -- nc -zv <hsm_server_ip> <hsm_port>
----

> . Monitor HSM server logs for authentication or capacity issues

Reviewer: ```suggestion
. Monitor HSM server logs for authentication or capacity issues.
```

== Getting additional help

If issues persist after following this troubleshooting guide:

. Collect adoption logs and configuration for analysis
. Check the HSM vendor documentation for vendor-specific troubleshooting
. Verify HSM server status and connectivity independently
. Review the adoption summary report for additional diagnostic information

> For Proteccio-specific issues, consult the Eviden Trustway documentation and ensure HSM server configuration matches the client settings.

Reviewer: What do you thin about moving this statement to the Proteccio HSM troubleshooting file? e.g.,
"If Proteccio HSM issues persist, consult the Eviden Trustway documentation and ensure HSM server configuration matches the client settings."
